h2o-core/src/main/java/water/jdbc/SQLManager.java

```diff
@@ -304,6 +341,10 @@ private static final int calculateLocalConnectionCount(final int maxTotalConnect
       //Make sure at least some connections are available to a node
       return Math.max(conPerNode, MIN_CONNECTIONS_PER_NODE);
     }
+
+    static int getOptimalConnectionPerNode(final int cloudSize, final short nThreads) {
+      return Math.min(nThreads, Math.max(getMaxConnectionsTotal() / cloudSize, MIN_CONNECTIONS_PER_NODE));
```

Comment:
I don't see the optimization here. The `Math.max(getMaxConnectionsTotal() / cloudSize, MIN_CONNECTIONS_PER_NODE)`. In many cases, this is similar to what I already have. In some edge cases, it may LOWER the number of connections (falling back to one connection per threads if the calculations result in higher number of connections than there are threads).

I'd image there is a reason behind having one connection per thread, yet some connections may just be in a state of waiting for resources and one thread can actually handle more than one connection while waiting for the others to be in the "ready" state. Is HIVE the reason for this optimization ?

Reply:
the point of this method is not to calculate the real number of connections used later: it is to make an estimation of the optimal amount of connections needed to retrieve the data, regardless of the size of the data...
The fact is that this "optimized" logic always use more connections than the previous one: the previous logic was creating 1 single chunk (and for this using 1 connection) for every 256MB of data. The current logic is trying to increase parallelism, but also avoiding waiting connections that would have a bad impact on fetching performance.
Basically most of the time consumed when fetching data os in iterating through the resultset, so waiting connections are useless, you just want to have as many parallel active connections as you can sucking up the data, without also having to make too many queries.

Reply:
let's discuss if more clarification needed

Reply:
closed after discussion